### PR TITLE
Split extensions into multiple modules

### PR DESCRIPTION
**Pull request by fishcakez**
Currently the decode method is chosen by getting the type information for oids and then finding an extension module that can handle the oid information. The oid is then mapped to that extension module. Then when it comes to decode a row every term (by default) is decoded by the same function, which means we have to parse the type information to choose the correct decode method.

This branch optimises this approach by splitting the built in extensions into many modules. This means no or minimal type information needs to be parsed.

As an example imagine we need to decode an "int4send" term with oid 1. Currently we lookup the type information and extension module for oid 1. The extension is `Postgrex.Extensions.Binary`. However this module decodes "everything" and so first it has to find the correct decode method by checking the type information. However because we know oid 1 is "int4send" we could use an extension that only handles "int4send". This module would only handle `int4send` and so does not need to parse the type information for every term.

To test the difference a simple benchmark for the difference in decode time can be done by putting the below in `test/bench_test.exs` and running `mix test test/bench_test.exs`
```elixir
defmodule BenchTest do
  use ExUnit.Case, async: true
  import Postgrex.TestHelper
  alias Postgrex.Connection, as: P

  setup do
    opts = [ database: "postgrex_test" ]
    {:ok, pid} = P.start_link(opts)
    {:ok, [pid: pid]}
  end

  test "benchmark simple decode", context do
    for n <- 1..10000 do
      :ok = query("insert into uniques values ($1)", [n])
    end
    %Postgrex.Result{} = res = P.query!(context[:pid], "SELECT * from uniques", [], [decode: :manual])
    n = 1000
    times = for _ <- 1..n do
      {time, _} = :timer.tc(Postgrex.Result, :decode, [res])
      time
    end
    IO.inspect div(Enum.sum(times), n)
  end
end
```

Locally this branch reduces decoding time by a third compared to master using that benchmark.

Currently 4 tests fail because the error handling isn't what it used to be for encoding. However because we split the types into multiple modules we might be able to provide better errors. For example when an "int4send" term is invalid we can state the integer range in the error.

### CODE REVIEW

**Comment by fishcakez** [comment]
tl:dr Pattern matching is not "free" and beam is very fast at decoding ints/floats/strings from binaries.

To understand what beam is doing we can create a beam assembly file from a module with this 1 liner (`module` must be bound to the desired module, e.g. `module = Postgrex.Extensions.Binary`):
```elixir
{:ok, {_, [{_, {_, ac}}]}} = :beam_lib.chunks(:code.which(module), [:abstract_code]) ; {:ok, _, _} = :compile.forms(ac, [:S]) ; "#{module}.S"
```
Now to go to `{function, decode, 4,` to find the decode function of arity 4:
```asm
{function, decode, 4, 104}.
  {label,103}.
    {line,[{location,"lib/postgrex/extensions/binary.ex",410}]}.
    {func_info,{atom,'Elixir.Postgrex.Extensions.Binary'},{atom,decode},4}.
  {label,104}.
    {test,is_map,{f,103},[{x,0}]}.
    {get_map_elements,{f,150},
                      {x,0},
                      {list,[{atom,'__struct__'},{x,5},{atom,send},{x,4}]}}.
    {test,is_eq_exact,{f,150},[{x,5},{atom,'Elixir.Postgrex.TypeInfo'}]}.
    {test,bs_start_match2,{f,150},6,[{x,4},0],{x,6}}.
    {test,bs_get_integer2,
          {f,150},
          7,
          [{x,6},
           {integer,8},
           1,
           {field_flags,[{anno,[410,
                                {file,"lib/postgrex/extensions/binary.ex"}]},
                         unsigned,big]}],
          {x,7}}.
    {test,is_integer,{f,150},[{x,7}]}.
    {select_val,{x,7},
                {f,150},
                {list,[{integer,117},
                       {f,105},
                       {integer,116},
                       {f,108},
                       {integer,98},
                       {f,116},
                       {integer,118},
                       {f,122},
                       {integer,110},
                       {f,125},
                       {integer,97},
                       {f,126},
                       {integer,105},
                       {f,127},
                       {integer,101},
                       {f,132},
                       {integer,102},
                       {f,133},
                       {integer,114},
                       {f,148},
                       {integer,100},
                       {f,149}]}}.
...
 {label,127}.
    {test,bs_match_string,{f,150},[{x,6},16,{string,"nt"}]}.
    {test,bs_get_integer2,
          {f,150},
          8,
          [{x,6},
           {integer,8},
           1,
           {field_flags,[{anno,[428,
                                {file,"lib/postgrex/extensions/binary.ex"}]},
                         unsigned,big]}],
          {x,8}}.
    {test,is_integer,{f,150},[{x,8}]}.
    {select_val,{x,8},
                {f,150},
                {list,[{integer,52},
                       {f,128},
                       {integer,56},
                       {f,129},
                       {integer,50},
                       {f,130},
                       {integer,101},
                       {f,131}]}}.
  {label,128}.
    {test,bs_match_string,{f,150},[{x,6},32,{string,"send"}]}.
    {test,bs_test_tail2,{f,150},[{x,6},0]}.
    {test,bs_start_match2,{f,150},9,[{x,1},0],{x,9}}.
    {test,bs_get_integer2,
          {f,150},
          10,
          [{x,9},
           {integer,32},
           1,
           {field_flags,[{anno,[430,
                                {file,"lib/postgrex/extensions/binary.ex"}]},
                         signed,big]}],
          {x,10}}.
    {test,bs_test_tail2,{f,150},[{x,9},0]}.
    {move,{x,10},{x,0}}.
    return.
```
This is the code for `decode/4`. This is what is happen for "int4send":
* Check first argument is map
* Get `:__struct__` and `:send` from map
* Check `:__struct__` value is `Postgrex.TypeInfo`
* Start binary match on `:send` value
* Get integer value of first byte of `:send` value
* Compare that integer value against possible values (`105 = ?i` for i in "int4send")
* Check 2 byte string for "nt"
* Check integer value for (`52 = ?4`)
* Check remaining binary is "send"
* Start binary match on 2nd argument.
* Get 4 byte integer.

Now on this branch it can go straight to matching on 2nd argument
```asm
{function, decode, 4, 9}.
  {label,8}.
    {line,[{location,"lib/postgrex/extensions/binary/int4.ex",11}]}.
    {func_info,{atom,'Elixir.Postgrex.Extensions.Binary.Int4'},
               {atom,decode},
               4}.
  {label,9}.
    {test,bs_start_match2,{f,8},4,[{x,1},0],{x,1}}.
    {test,bs_get_integer2,
          {f,10},
          4,
          [{x,1},
           {integer,32},
           1,
           {field_flags,[{anno,[11,
                                {file,"lib/postgrex/extensions/binary/int4.ex"}]},
                         signed,big]}],
          {x,4}}.
    {test,bs_test_tail2,{f,10},[{x,1},0]}.
    {move,{x,4},{x,0}}.
    return.
```
A smaller optimisation could be done, which is to reduce the matching on "int4send" by replacing it with the atom `:int4send`. A match against an atom is equivalent to a match against an integer.

**Comment by josevalim** [comment]
@ericmj do you remember what is holding us back on this one? I want to do an optimization round before Ecto 2.0.

**Comment by ericmj** [comment]
I don't like the approach of splitting into multiple modules, it seems unnecessary. There are multiple modules but still only a single extension. Why not just split into multiple functions?

**Comment by fishcakez** [comment]
If we can break API I am happy to redo this with multiple functions.

**Comment by ericmj** [comment]
What would the breaking API be?

**Comment by fishcakez** [comment]
We would have to change the extension format to reference a function along with the encoding.

**Comment by ericmj** [comment]
Why? Why can't we change this line https://github.com/fishcakez/postgrex/blob/extension_split/lib/postgrex/extensions/binary.ex#L66 to dispatch to a local function instead of a remote?

**Comment by fishcakez** [comment]
Oh that module is there because I was trying to allow people to use that extension for backwards compatibility. To take avdvantage of the optimisation you use the modules as separate extensions and dont use binary extension.

**Comment by ericmj** [comment]
The binary extension is private, I just saw that by mistake text is public. Maybe the extensions should be public but they are not intended to be right now so we can break them if we want. I am also not sure what we would be breaking since the public API would be the same (it still conforms to the extension behaviour even in the current PR).

**Comment by fishcakez** [comment]
Oh I thought it was public because of the failing tests, I thought the failing tests meant API backwards incompatible.

**Comment by fishcakez** [comment]
Oh nevermind not failing tests. It was related to this: https://github.com/ericmj/postgrex/blob/master/test/custom_extensions_test.exs#L20. As using the binary extension would not work there anymore.

**Comment by ericmj** [comment]
The test tests if you can nest extension calls, it just uses the binary extension because it was convenient. Maybe they should be public but they are not now so it's fine to change. I would rather change the extension API to map from type/sender to anon function but I feel like that complicates the API and you wouldn't be able to do what the linked test does if we do decide to make the extensions public.

What are the performance gains from this PR?

**Comment by fishcakez** [comment]
> Locally this branch reduces decoding time by a third compared to master using that benchmark.

**Comment by ericmj** [comment]
With the following benchmark where we measure the performance increase in query time instead of just decode time we just see 2.3% performance improvement when measuring against a local postgres server.

Because this change will complicate the internal code, public API and adds almost 200 LOC I don't think it's worth it unless we can simplify it.

```elixir
defmodule BenchTest do
  use ExUnit.Case, async: true
  import Postgrex.TestHelper
  alias Postgrex.Connection, as: P

  setup do
    opts = [ database: "postgrex_test" ]
    {:ok, pid} = P.start_link(opts)
    {:ok, [pid: pid]}
  end

  test "benchmark simple decode", context do
    for n <- 1..10000 do
      :ok = query("insert into uniques values ($1)", [n])
    end

    n = 1000
    times = for _ <- 1..n do
      {time, _} = :timer.tc(fn ->
        %Postgrex.Result{} = P.query!(context[:pid], "SELECT * from uniques", [], [decode: :manual])
      end)
      time
    end

    IO.inspect div(Enum.sum(times), n)
  end
end
```

**Comment by fishcakez** [comment]
Can you remove the `decode: :manual` and try again because that snippet does not do any decoding so peformance should be the same. This branch reduces the total query time by 10% for me.

We can remove ~85 lines from Binary.Extension if it is private so this is a 10% reduction for only ~120 lines.

**Comment by fishcakez** [comment]
I've rebased on master and cleaned things up a bit. There some failing tests still as argument errors are not raised when extension cant encode. If extensions are split into multiple modules we could give detailed information about what type is expected, rather than a generic message.

**Comment by ericmj** [comment]
I see a 6% improvement but I haven't tested your cleaned up branch yet. We can improve encode errors today by printing information from TypeInfo when we have it. Could we improve it additionally with this PR?

**Comment by fishcakez** [comment]
No, rather it would be silly not to do with this PR. But we could do it without this PR.

**Comment by ericmj** [comment]
> No, rather it would be silly not to do with this PR. But we could do it without this PR.

Makes sense.

I'm leaning towards including this because the benefits are clear now. If we can find a nice way of still supporting a performant variant of this https://github.com/ericmj/postgrex/blob/master/test/custom_extensions_test.exs#L19-L23 it would be great but it's not blocking this PR. In an early version of extensions you could recursively call decode but it would skip the current extension but I believe it was scrapped when we started caching types. Maybe we can implement something like that again?

File: lib/postgrex/extensions/network.ex [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  alias Postgrex.TypeInfo
+  use Postgrex.BinaryExtension,
+    [send: "inet_send", send: "cidr_send", send: "macaddr_send"]
+
+  def encode(%TypeInfo{type: "inet"}, %Postgrex.INET{} = inet, _, _),
+    do: encode_network(inet)
+  def encode(%TypeInfo{type: "cidr"}, %Postgrex.CIDR{} = cidr, _, _),
+    do: encode_network(cidr)
+  def encode(%TypeInfo{type: "macaddr"}, %Postgrex.MACADDR{} = macaddr, _, _),
+    do: encode_network(macaddr)
```
Comment (ericmj) [inline lib/postgrex/extensions/network.ex]: We still seem to have a few extensions that still match on `TypeInfo`. Would it be worth to split them out and should we map the oid to anonymous functions instead of modules in that case?

Comment (fishcakez) [inline lib/postgrex/extensions/network.ex]: This example is for encode, not decode, so didn't think was worth it as encoding is usually a few parameters and decoding is many rows. The decode code does not use type info.

Great you mentioned this though, macaddr technically needs to be moved to is own extension because it could conflicts with an inet ipv4 address.

File: lib/postgrex/extensions/pg_extension.ex [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  alias Postgrex.TypeInfo
+  import Postgrex.BinaryUtils
+  use Postgrex.BinaryExtension, type: "citext", type: "hstore"
+
+  def encode(%TypeInfo{type: "citext"}, bin, _, _) when is_binary(bin),
+    do: bin
+  def encode(%TypeInfo{type: "hstore"}, map, _, _),
+    do: encode_hstore(map)
+
+  def decode(%TypeInfo{type: "citext"}, bin, _, _),
```
Comment (fishcakez) [inline lib/postgrex/extensions/pg_extension.ex]: I forget why `citext` is here.. it should be moved to raw because it doesnt need any encode/decde and then we dont use type info. This extension would then be renamed hstore.

**Comment by fishcakez** [comment]
> Maybe we can implement something like that again?

We don't array data, the lookup is done in the ets table. Currently we lookup entries by `oid` integer key, we could also add every extension listed as {module, oid} and do the `module` based lookup on the ets table.

**Comment by fishcakez** [comment]
> I believe it was scrapped when we started caching types

It was not scrapped it still works, it just doesnt use the cache.

**Comment by fishcakez** [comment]
Updated.

I did not solve the decode using a specific module extension issue.

**Comment by fishcakez** [comment]
I've added 2 new commits, please check these and I think we are good to go.
